serving/features.go

```diff
@@ -0,0 +1,284 @@
+			return true
+		})
+		serv.Features.Store(serv.getNVCacheKey(name, variant), metaFeature)
+		return metaFeature, nil
+	}
+}
+
+func (serv *FeatureServer) getPrecomputedValues(ctx context.Context, entityMap map[string][]string, meta *metadata.FeatureVariant) ([]indexedValue, error) {
+	logger := serv.Logger
+	obs := ctx.Value(observer{}).(metrics.FeatureObserver)
+	entities, has := entityMap[meta.Entity()]
+	if !has {
+		logger.Errorw("Entity not found", "Entity", meta.Entity())
+		obs.SetError()
+		return nil, fmt.Errorf("no value for entity %s", meta.Entity())
+	}
+
+	store, err := serv.cacheFeatureProvider(ctx, meta)
+	if err != nil {
+		logger.Errorw("Could not fetch provider", "Entity", meta.Entity())
+		obs.SetError()
+		return nil, fmt.Errorf("could not fetch online provider %s", meta.Provider())
+	}
+
+	featureTable, err := serv.cacheFeatureTable(ctx, store, meta.Name(), meta.Variant())
+	if err != nil {
+		return nil, err
+	}
+
+	featureValues, err := serv.getEntityValues(ctx, entities, featureTable)
+	if err != nil {
+		return nil, err
+	}
+	return featureValues, nil
+
+}
+
+func (serv *FeatureServer) cacheFeatureProvider(ctx context.Context, meta *metadata.FeatureVariant) (provider.OnlineStore, error) {
+	if store, has := serv.Providers.Load(meta.Provider()); has {
+		return store.(provider.OnlineStore), nil
+	} else {
+		store, err := serv.initializeFeatureProvider(ctx, meta)
+		if err != nil {
+			return nil, err
+		}
+		serv.Providers.Store(meta.Provider(), store)
+		return store.(provider.OnlineStore), nil
+	}
+}
+
+func (serv *FeatureServer) initializeFeatureProvider(ctx context.Context, meta *metadata.FeatureVariant) (provider.OnlineStore, error) {
+	logger := serv.Logger
+	obs := ctx.Value(observer{}).(metrics.FeatureObserver)
+	providerEntry, err := meta.FetchProvider(serv.Metadata, ctx)
+	if err != nil {
+		logger.Errorw("fetching provider metadata failed", "Error", err)
+		obs.SetError()
+		return nil, fmt.Errorf("fetching provider metadata failed: %w", err)
+	}
+	p, err := provider.Get(pt.Type(providerEntry.Type()), providerEntry.SerializedConfig())
+	if err != nil {
+		logger.Errorw("failed to get provider", "Error", err)
+		obs.SetError()
+		return nil, fmt.Errorf("failed to get provider: %w", err)
+	}
+	store, err := p.AsOnlineStore()
+	if err != nil {
+		logger.Errorw("failed to use provider as onlinestore for feature", "Error", err)
+		obs.SetError()
+		// This means that the provider of the feature isn't an online store.
+		// That shouldn't be possible.
+		return nil, fmt.Errorf("failed to use provider as onlinestore for feature: %w", err)
+	}
+	return store, nil
+}
+
+func (serv *FeatureServer) cacheFeatureTable(ctx context.Context, store provider.OnlineStore, name, variant string) (provider.OnlineStoreTable, error) {
+	obs := ctx.Value(observer{}).(metrics.FeatureObserver)
+
+	var featureTable provider.OnlineStoreTable
+	if table, has := serv.Tables.Load(serv.getNVCacheKey(name, variant)); has {
```

Comment:
does the cache expire?

Reply:
I wonder if caching the provider table details is something we should always do (regardless of features)

Reply:
Cache never expires since tables should never change (at the moment). We should probably have a check to make sure the table is valid and try to refresh, but don't currently have the tooling to do that.